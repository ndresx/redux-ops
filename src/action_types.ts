export const prefix = '@@redux-ops/';

export const START = `${prefix}START`;
export const UPDATE = `${prefix}UPDATE`;
export const DELETE = `${prefix}DELETE`;
export const CLEAR = `${prefix}CLEAR`;

export const BLUEPRINT = `${prefix}BLUEPRINT`;
