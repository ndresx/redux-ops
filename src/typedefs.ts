import { AnyAction, Action } from 'redux';

export interface Operation<TData = any> {
  readonly id: OpId;
  readonly status: OpStatus;
  readonly data?: TData;
}

export type OpId = number | string;

export enum OpStatus {
  Start = 'start',
  Delete = 'delete',
  Success = 'success',
  Error = 'error',
}

// State
export interface Ops<TData = any> {
  readonly [key: string]: Operation<TData>;
}

export interface OpsState extends Ops {}

export type OpsReducerHandler<A extends AnyAction = AnyAction> = (
  state: OpsState,
  action: A
) => OpsState;

export type OperationActionHandler = (state: OpsState, action: OperationAction) => OpsState;

// Actions
export interface OperationAction<TData = any> extends Action {
  readonly payload: Operation<TData>;
}

export interface ResetOperationsAction extends AnyAction {}

export interface OperationActionCreator {
  <TData = any>(id: OpId, status?: OpStatus, data?: TData): OperationAction<TData>;
}

export interface DeleteOperationActionCreator {
  (id: OpId): OperationAction;
}

export interface ResetOperationsActionCreator {
  (): AnyAction;
}
