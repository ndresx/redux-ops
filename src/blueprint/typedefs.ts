import { AnyAction } from 'redux';

import * as actionTypes from '../action_types';
import { OperationAction, DeleteOperationAction, OpId } from '../typedefs';

export enum BlueprintActionKey {
  Start = 'start',
  Success = 'success',
  Error = 'error',
  Delete = 'delete',
}

export type BlueprintActionTypes = {
  readonly STARTED: string;
  readonly INTERMEDIATE: string;
  readonly SUCCESS: string;
  readonly ERROR: string;
};

export interface OpBlueprintAction {
  readonly type: string;
  [actionTypes.prefix]: OpsData;
}

export interface OpsData extends OpsGenericActionData {
  op: OperationAction | DeleteOperationAction;
  action?: OpBlueprintOriginalAction | null;
}

export interface OpsGenericActionData {
  uniqueId?: OpId;
  broadcast?: boolean;
}

export interface OpBlueprintOriginalAction extends AnyAction {
  [actionTypes.prefix]?: OpsGenericActionData;
}

export interface BlueprintActionCreator {
  (data?: any, action?: OpBlueprintOriginalAction | null): OpBlueprintAction;
}

type OpBlueprintActionCreator = (...args: any[]) => any;

export type BlueprintActionComposers = { [key in BlueprintActionKey]?: OpBlueprintActionCreator };

export type ComposedActionCreator = BlueprintActionCreator | OpBlueprintActionCreator;

export type OpBlueprint = { [key in BlueprintActionKey]: ComposedActionCreator } & {
  readonly id: OpId;
};

export type OpBlueprintFn<T extends (...args: any) => any> = (
  ...args: Parameters<T>
) => OpBlueprintAction;
