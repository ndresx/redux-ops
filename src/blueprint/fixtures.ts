import { AnyAction } from 'redux';
import { OpBlueprint, OpBlueprintFn } from './typedefs';

export interface MovieFetcherOp extends OpBlueprint {
  readonly start: OpBlueprintFn<typeof fetchMovies>;
  readonly success: OpBlueprintFn<typeof didFetchMovies>;
}

export const movies = [
  { id: 1, name: 'The Dark Knight' },
  { id: 2, name: 'Jurassic World' },
  { id: 3, name: 'Avatar' },
];

export const FETCH_MOVIES = 'FETCH_MOVIES';

export function fetchMovies(genre: string, page?: number): AnyAction {
  return {
    type: FETCH_MOVIES,
    payload: { genre, page },
  };
}

export function didFetchMovies(movies: object): AnyAction {
  return {
    type: `${FETCH_MOVIES}_SUCCESS_CUSTOM`,
    payload: { movies },
  };
}
