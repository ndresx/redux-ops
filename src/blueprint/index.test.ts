import {
  createBlueprint,
  opUnique,
  opBroadcast,
  getUniqueId,
  createBlueprintActionTypes,
} from './index';
import { OpBlueprint, BlueprintActionKey } from './typedefs';
import { fetchMovies, FETCH_MOVIES, movies, didFetchMovies } from './fixtures';
import { prefix } from '../action_types';

describe('blueprint', () => {
  function testActions(
    blueprint: OpBlueprint,
    blueprintArguments: { [key in BlueprintActionKey]?: unknown[] } = {}
  ): void {
    expect(blueprint).toMatchSnapshot();

    for (const key in BlueprintActionKey) {
      it(key, () => {
        const keyValue = BlueprintActionKey[key];
        expect(blueprint[keyValue](...(blueprintArguments[keyValue] || []))).toMatchSnapshot();
      });
    }
  }

  describe('createBlueprint', () => {
    describe('should create blueprint with default actions', () => {
      testActions(createBlueprint(FETCH_MOVIES));
    });

    describe('should create blueprint with default actions and custom payloads', () => {
      testActions(createBlueprint(FETCH_MOVIES), {
        start: ['Science-Fiction', 1],
        success: [movies],
        error: [new Error('404').message],
      });
    });

    describe('should create blueprint with custom actions', () => {
      testActions(
        createBlueprint(FETCH_MOVIES, {
          start: fetchMovies,
          success: didFetchMovies,
        }),
        {
          start: ['Science-Fiction', 1],
          success: [movies],
          error: [new Error('404').message],
        }
      );
    });

    it('should create blueprint action types', () => {
      expect(createBlueprintActionTypes(FETCH_MOVIES)).toMatchSnapshot();
    });
  });

  describe('opUnique', () => {
    let blueprintModule;

    beforeEach(() => {
      blueprintModule = require('./index');
    });

    it('should assign unique id(s) to blueprint action', () => {
      const blueprint = blueprintModule.createBlueprint(FETCH_MOVIES);
      const startAction = blueprint.start();

      const firstStartAction = blueprintModule.opUnique(startAction);
      expect(firstStartAction[prefix].uniqueId).toMatchInlineSnapshot(
        `"@@redux-ops/FETCH_MOVIES_1"`
      );

      const secondStartAction = blueprintModule.opUnique(startAction);
      expect(secondStartAction[prefix].uniqueId).toMatchInlineSnapshot(
        `"@@redux-ops/FETCH_MOVIES_2"`
      );
    });

    it('should re-assign unique id to existing blueprint action', () => {
      const blueprint = blueprintModule.createBlueprint(FETCH_MOVIES);
      const startAction = blueprint.start();

      const firstStartAction = blueprintModule.opUnique(startAction);
      expect(firstStartAction[prefix].uniqueId).toMatchInlineSnapshot(
        `"@@redux-ops/FETCH_MOVIES_1"`
      );

      const secondStartAction = blueprintModule.opUnique(startAction, firstStartAction);
      expect(secondStartAction[prefix].uniqueId).toMatchInlineSnapshot(
        `"@@redux-ops/FETCH_MOVIES_1"`
      );
    });

    it('should assign custom unique id to blueprint action', () => {
      const blueprint = blueprintModule.createBlueprint(FETCH_MOVIES);
      const startAction = blueprint.start();

      const firstStartAction = blueprintModule.opUnique(startAction, '123');
      expect(firstStartAction[prefix].uniqueId).toMatchInlineSnapshot(`"123"`);

      const secondStartAction = blueprintModule.opUnique(startAction, firstStartAction);
      expect(secondStartAction[prefix].uniqueId).toMatchInlineSnapshot(`"123"`);
    });

    it('should assign custom unique id to custom blueprint action', () => {
      const blueprint = blueprintModule.createBlueprint(FETCH_MOVIES, {
        start: fetchMovies,
      });
      const startAction = blueprint.start();

      expect(startAction[prefix].action!.uniqueId).toBeUndefined();
      expect(opUnique(startAction)[prefix].action!.uniqueId).toMatchInlineSnapshot(`undefined`);
    });

    it('should get unique id of blueprint action', () => {
      const blueprint = blueprintModule.createBlueprint(FETCH_MOVIES);
      const startAction = blueprint.start();

      expect(getUniqueId(startAction)).toBeUndefined();
      expect(getUniqueId(blueprintModule.opUnique(startAction))).toMatchInlineSnapshot(
        `"@@redux-ops/FETCH_MOVIES_1"`
      );
    });

    it('should create blueprint with actions using unique ids', () => {
      const module: typeof import('./index') = require('./index');
      const blueprint = module.opUnique(module.createBlueprint(FETCH_MOVIES));
      const blueprintCustomId = module.opUnique(module.createBlueprint(FETCH_MOVIES), '123');

      for (const key in BlueprintActionKey) {
        const keyValue = BlueprintActionKey[key];
        expect(blueprint[keyValue]()[prefix].uniqueId).toBe('@@redux-ops/FETCH_MOVIES_1');
        expect(blueprintCustomId[keyValue]()[prefix].uniqueId).toBe('123');
      }

      jest.resetModules();
    });

    afterEach(() => {
      jest.resetModules();
    });
  });

  describe('opBroadcast', () => {
    it('should create broadcast blueprint action', () => {
      const blueprint = createBlueprint(FETCH_MOVIES);
      const startAction = blueprint.start();

      expect(startAction[prefix].broadcast).toBeUndefined();
      expect(opBroadcast(startAction)[prefix].broadcast).toBe(true);
    });

    it('should create blueprint with broadcast actions', () => {
      const module: typeof import('./index') = require('./index');
      const blueprint = module.opBroadcast(module.createBlueprint(FETCH_MOVIES));

      for (const key in BlueprintActionKey) {
        const keyValue = BlueprintActionKey[key];
        expect(blueprint[keyValue]()[prefix].broadcast).toBe(true);
      }

      jest.resetModules();
    });
  });
});
