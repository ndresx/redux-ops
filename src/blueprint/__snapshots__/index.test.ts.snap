// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "delete": [Function],
  "error": [Function],
  "id": "FETCH_MOVIES",
  "start": [Function],
  "success": [Function],
}
`;

exports[` 2`] = `
Object {
  "delete": [Function],
  "error": [Function],
  "id": "FETCH_MOVIES",
  "start": [Function],
  "success": [Function],
}
`;

exports[` 3`] = `
Object {
  "delete": [Function],
  "error": [Function],
  "id": "FETCH_MOVIES",
  "start": [Function],
  "success": [Function],
}
`;

exports[`blueprint createBlueprint should create blueprint action types 1`] = `
Object {
  "ERROR": "FETCH_MOVIES_ERROR",
  "INTERMEDIATE": "FETCH_MOVIES_INTERMEDIATE",
  "STARTED": "FETCH_MOVIES_STARTED",
  "SUCCESS": "FETCH_MOVIES_SUCCESS",
}
`;

exports[`blueprint createBlueprint should create blueprint with custom actions Delete 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "id": "FETCH_MOVIES",
      },
      "type": "@@redux-ops/DELETE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with custom actions Error 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": "404",
        "id": "FETCH_MOVIES",
        "status": "error",
      },
      "type": "@@redux-ops/UPDATE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with custom actions Start 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": Object {
      "payload": Object {
        "genre": "Science-Fiction",
        "page": 1,
      },
      "type": "FETCH_MOVIES",
    },
    "op": Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "started",
      },
      "type": "@@redux-ops/START",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with custom actions Success 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": Object {
      "payload": Object {
        "movies": Array [
          Object {
            "id": 1,
            "name": "The Dark Knight",
          },
          Object {
            "id": 2,
            "name": "Jurassic World",
          },
          Object {
            "id": 3,
            "name": "Avatar",
          },
        ],
      },
      "type": "FETCH_MOVIES_SUCCESS_CUSTOM",
    },
    "op": Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "success",
      },
      "type": "@@redux-ops/UPDATE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions Delete 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "id": "FETCH_MOVIES",
      },
      "type": "@@redux-ops/DELETE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions Error 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "error",
      },
      "type": "@@redux-ops/UPDATE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions Start 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "started",
      },
      "type": "@@redux-ops/START",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions Success 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": undefined,
        "id": "FETCH_MOVIES",
        "status": "success",
      },
      "type": "@@redux-ops/UPDATE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions and custom payloads Delete 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "id": "FETCH_MOVIES",
      },
      "type": "@@redux-ops/DELETE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions and custom payloads Error 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": "404",
        "id": "FETCH_MOVIES",
        "status": "error",
      },
      "type": "@@redux-ops/UPDATE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions and custom payloads Start 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": "Science-Fiction",
        "id": "FETCH_MOVIES",
        "status": "started",
      },
      "type": "@@redux-ops/START",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;

exports[`blueprint createBlueprint should create blueprint with default actions and custom payloads Success 1`] = `
Object {
  "@@redux-ops/": Object {
    "action": undefined,
    "op": Object {
      "payload": Object {
        "data": Array [
          Object {
            "id": 1,
            "name": "The Dark Knight",
          },
          Object {
            "id": 2,
            "name": "Jurassic World",
          },
          Object {
            "id": 3,
            "name": "Avatar",
          },
        ],
        "id": "FETCH_MOVIES",
        "status": "success",
      },
      "type": "@@redux-ops/UPDATE",
    },
  },
  "type": "@@redux-ops/BLUEPRINT",
}
`;
